### ISSUE TITLE
Software licence

### ISSUE THREAD

**Issue by aidanheerdegen** [COSIMA/access-om2#264]
There is currently no licence for the ACCESS-OM2 model. I don't know if this is a problem or not, but currently looking into licensing for the ACCESS-NRI and noticed the absence of a licence.

Note that there are some licence incompatibility issues with MOM5

https://github.com/mom-ocean/MOM5/issues/370

**Comment by aekiss** [comment]
There are lots ACCESS-OM2 components to check for license compatibility...
![Screen Shot 2022-08-16 at Tue 16-8 3 31pm](https://user-images.githubusercontent.com/31054815/184804848-ff13831a-d833-4870-b3f5-366a7eb721b6.png)

**Comment by aidanheerdegen** [comment]
CODE COMPONENT SPAGHETTI!

Yeah I get it. I'm all for the "Ostrich Licence" approach, but does that even work if components have licences? 

I really wish I didn't have to care about this.

**Comment by micaeljtoliveira** [comment]
Looks like you have a serious issue here with MOM5... 

Also keep in mind that any code that is in this repository, but not on any submodules, like tests, utilities, etc also needs to have a license (and each file needs a copyright notice).

**Comment by aidanheerdegen** [comment]
Just so @aekiss doesn't feel too burdened, it is a *recommendation* that code files include the copyright statement, it isn't compulsory.

**Comment by aekiss** [comment]
So let's skip it then

**Comment by aidanheerdegen** [comment]
Ostrich is off the table. I repeat, ostrich is off the table

https://opensource.guide/legal/#are-public-github-projects-open-source

> Making your GitHub project public is not the same as licensing your project. Public projects are covered by [GitHub’s Terms of Service](https://help.github.com/en/github/site-policy/github-terms-of-service#3-ownership-of-content-right-to-post-and-license-grants), which allows others to view and fork your project, but your work otherwise comes with no permissions.
> If you want others to use, distribute, modify, or contribute back to your project, you need to include an open source license. For example, someone cannot legally use any part of your GitHub project in their code, even if it’s public, unless you explicitly give them the right to do so.

That isn't the only source I've found for the assertion that no license means no safe re-use. Now this is all theoretical, I certainly don't think you're going to get sued by anyone, but patent trolls are real, and the complicated network of copyright and patent portfolios (c.f. SCO Unix) means that it is not impossible.

**Comment by aekiss** [comment]
too late!
![](https://images.vice.com/vice/images/galleries/meta/2014/03/19/148283-1413265360820.jpeg)

**Comment by aekiss** [comment]
But seriously, what would be the most widely-compatible license?

**Comment by aidanheerdegen** [comment]
Check this from CESM

https://github.com/ESCOMP/CESM/blob/master/LICENSE.txt

**Comment by aekiss** [comment]
hm, OK so that's a DIY license -- would it be better to use something standard?

**Comment by aidanheerdegen** [comment]
> But seriously, what would be the most widely-compatible license?

MIT is very open. Ditto apache. Essentially a free-for-all, knock yourself out. Check out licences here:

https://choosealicense.com

But this stuff is horrible, I hate it, I wish I wasn't discussing it. If you include GPL code into your own code, then you're obligated to have a GPL licence, or strictly denote which bits are GPL and which are, say, MIT or apache. BUT I AM NOT A LAWYER SO THIS IS JUST MY RECKONS FROM READING SOME STUFF ON THE INTERNET.

**Comment by aekiss** [comment]
We have some conditions of use here: https://github.com/COSIMA/access-om2#conditions-of-use
so maybe those could be extended to include disclaimers, copyright, terms for distribution, etc?

**Comment by aidanheerdegen** [comment]
Interestingly GPL licences explicitly *don't* allow for binding someone who uses the code to quote/cite it. Only to reproduce the licence when making derivative works. So stuff like that is more of a request than an obligation.

**Comment by aekiss** [comment]
going through the horrendogram [above](https://github.com/COSIMA/access-om2/issues/264#issuecomment-1216168593), here's an overview of the licensing of the various components:
- [MOM5](https://github.com/mom-ocean/MOM5) - currently [GPLv2](https://github.com/mom-ocean/MOM5/blob/master/LICENSE) but we might change it, e.g. to GPLv3 or LGPLv3 for compatibility with FMS https://github.com/mom-ocean/MOM5/issues/370
    - [FMS](https://github.com/mom-ocean/MOM5/tree/master/src/shared/fms) subtree: uses [LGPLv3](https://github.com/mom-ocean/MOM5/blob/master/src/shared/fms/fms.F90)
- [our CICE5 fork](https://github.com/COSIMA/cice5) - currently none; change to [the CICE5 custom permissive license](https://github.com/CICE-Consortium/CICE-svn-trunk/blob/main/LICENSE.pdf)? https://github.com/COSIMA/cice5/issues/67
    - [ParallelIO](https://github.com/NCAR/ParallelIO) submodule: [Apache License 2.0](https://github.com/NCAR/ParallelIO/blob/master/LICENSE)
- [libaccessom2](https://github.com/COSIMA/libaccessom2): no license; would need to be something compatible with these submodules:
     - [our oasis3-mct v2 fork](https://github.com/COSIMA/oasis3-mct): no license given; should probably be LGPL to match what more recent versions use https://github.com/COSIMA/oasis3-mct/issues/20
         - [MCT](https://github.com/COSIMA/oasis3-mct/tree/master/lib/mct): a [custom permissive license](https://github.com/COSIMA/oasis3-mct/blob/master/lib/mct/COPYRIGHT)
         - [SCRIP](https://github.com/COSIMA/oasis3-mct/tree/master/lib/scrip/src): a [custom permissive license](https://github.com/COSIMA/oasis3-mct/blob/master/lib/scrip/src/scrip.F)
     - [datetime-fortran](https://github.com/wavebitscientific/datetime-fortran): uses the [MIT license](https://github.com/wavebitscientific/datetime-fortran/blob/main/LICENSE)
     - [json-fortran](https://github.com/jacobwilliams/json-fortran): a [custom permissive license](https://github.com/jacobwilliams/json-fortran/blob/master/LICENSE)
- tools/[esmgrids](https://github.com/COSIMA/esmgrids): [Apache-2.0 license](https://github.com/COSIMA/esmgrids/blob/master/LICENSE.md)

**Comment by micaeljtoliveira** [comment]
>  If you include GPL code into your own code, then you're obligated to have a GPL licence, or strictly denote which bits are GPL and which are, say, MIT or apache. 

The main issue you have here is that you are combining things that are not GPL with GPL code. That's a big no... You can use non-GPL **from** GPL, but usually not the way around. So you definitely need to relicense MOM5, because although you can choose the license you want for the code in this repo and in some of the components (e.g,, libaccessom2), you are not the copyright holder of some of the components that have an incompatible license.

**Comment by micaeljtoliveira** [comment]
Concerning the choice of license for this repo and libaccessom2, any license that is compatible with the most restrictive license from the components is fine. Once you change the license of MOM5, you can basically choose any of the standard non-GPL open-source licenses. For libraries, I'm a big fan of the Mozilla Public License, because it provides you, as a developer, with almost all the protections of LGPL, while still allowing static linking with non-(L)GPL code.

**Comment by aidanheerdegen** [comment]
Any experience or expertise is welcome @micaeljtoliveira!

When using an LGPL licence it is common to assign copyright to the Free Software Foundation, so they can lawyer-up on your behalf. Seems a bit odd to me, and who has the copyright to the code developed in here? 

My head hurts.

Thanks for doing the hard yards on the licensing @aekiss. I am sorry that you have been pulled into this ball of misery and confusion. Turns out GPL is viral for despair as well.

**Comment by micaeljtoliveira** [comment]
> When using an LGPL licence it is common to assign copyright to the Free Software Foundation, so they can lawyer-up on your behalf.

No, that's not common. Usually is only done for GNU software. Note that you cannot assign copyright to the FSF just like that. You need every copyright holder to sign an assignment contract  and send it to the FSF (assuming they are actually interested in your code!)

>  who has the copyright to the code developed in here?

In general the copyright belongs to the person who wrote the code. It can also belong to your employer, depending on the conditions of your employment contract.

**Comment by aidanheerdegen** [comment]
Thanks @micaeljtoliveira, I clearly got myself quite confused. See, I told you I didn't know what I was talking about.

Looking at the contributors, it is only @aekiss, @nichannah and me (@aidanheerdegen) who have substantial contributions to this repo. If we can get @nichannah to give us agreement I think adding a license shouldn't be a problem. Just have to choose which one. 

Mozilla says it is explicitly *not* compatible with GPL

https://www.mozilla.org/en-US/MPL/license-policy/

> The following licenses are not compatible with the Mozilla Public License, version 2.0:
>
> CC-BY, CC-BY-*
> GPL

But this site disagrees

https://fossa.com/blog/open-source-software-licenses-101-mozilla-public-license-2-0/

> License compatibility: The MPL 2.0 is compatible with GPL v2 or later, LGPL 2.1 or later, and/or AGPL 3.0 or later. The CDDL (and earlier versions of the MPL) are not.

As does [this wikipedia article.](https://en.wikipedia.org/wiki/Comparison_of_free_and_open-source_software_licenses#Approvals

Perhaps the Mozilla page is referring explicitly to a version of the GPL?

**Comment by micaeljtoliveira** [comment]
@aidanheerdegen Weird. This other page from the Mozilla website says it's compatible (the whole page is about how to combine MPL2 and GPL!): https://www.mozilla.org/en-US/MPL/2.0/combining-mpl-and-gpl/

**Comment by aidanheerdegen** [comment]
IKR. I feel justified in my fear and loathing of licensing.

**Comment by aidanheerdegen** [comment]
An important decision to make: do you want copyleft or permissive? Because Mozilla is apparently weak copyleft:

https://fossa.com/blog/open-source-software-licenses-101-mozilla-public-license-2-0/

The NRI is looking like going for Apache-2.0 as it is permissive but modified versions that are distributed must include notification of the modifications, so it is clear who did what.

https://fossa.com/blog/open-source-licenses-101-apache-license-2-0/

**Comment by micaeljtoliveira** [comment]
The fact that the MPL2 is weak copyleft is precisely the reason why I like it, as it makes sure any redistributed modifications of your code are kept open. It's very annoying to learn that someone took your code, improved it and/or fixed bugs and then redistributed it in binary form without disclosing the modified sources. 

Personally, I don't see any disadvantage of the MP2 regarding the Apache 2. Since modifications are public with the MPL2, it's clear what was modified and by whom. On the other hand, the Apache 2 is incompatible with the GPL2, which might be a headache in the future (I recently discovered that I have this exact issue in a code that has two dependencies, one using the GPL2 and the other one the Apache 2...)

**Comment by aidanheerdegen** [comment]
Good point. Seeing as other major codes are copyleft seems sensible to keep on that side of the fence.

This will languish unless I keep banging on about it, so can we (I?) take a licence change proposal to the next COSIMA Meeting @aekiss? Offer up MPL2 as an explicit choice.

**Comment by aekiss** [comment]
I can add it to the agenda for this Thurs if you like @aidanheerdegen. It would require a short background presentation so that people have some grasp of these licensing issues.

**Comment by aidanheerdegen** [comment]
Yeah, I was afraid you'd say that. I'm happy to try, but as demonstrated in this thread, I am frequently wrong and particularly when it comes to this topic.

Happy for someone else more knowledgeable to do it ... *cough* ... @micaeljtoliveira ... *cough*

**Comment by micaeljtoliveira** [comment]
I don't mind doing it, but tomorrow I most likely won't be able to join the meeting.

**Comment by aidanheerdegen** [comment]
Thanks @micaeljtoliveira. Next week then?

**Comment by aidanheerdegen** [comment]
I was away for the COSIMA Meeting, did this get discussed? Was there any resolution @aekiss?

**Comment by aidanheerdegen** [comment]
This is a nice overview of licensing by the lead developer of [spack](spack.io) which specifically recommends `APACHE-2.0 WITH LLVM-Exception`

https://www.osti.gov/biblio/1782513

**Comment by aekiss** [comment]
Sorry @aidanheerdegen I didn't take notes and the specifics of that COSIMA meeting have slipped beyond my event horizon. @micaeljtoliveira do you recall?

**Comment by micaeljtoliveira** [comment]
In the end I don't think we discussed it. IIRC, we were supposed to do it on the 29 Sep., but I was sick that day.

**Comment by aekiss** [comment]
hm, and we forgot to discuss this at the last TWG too.... how about next one? Though I'll probably be away for that.

**Comment by micaeljtoliveira** [comment]
> how about next one? 

That would be a good idea. I believe it's better to discuss this during a TWG meeting, than during a weekly COSIMA meeting

> Though I'll probably be away for that.

Hmm, that would be unfortunate. When are you coming back?

**Comment by aekiss** [comment]
at the start of January

**Comment by micaeljtoliveira** [comment]
Then  maybe it's best to discuss this during the January TWG meeting.

**Comment by aekiss** [comment]
Or discuss in Dec without me? I have neither expertise nor strong opinions, so I don't feel I need to be present for a decision....

**Comment by aidanheerdegen** [comment]
I have asked @micaeljtoliveira to add this to the next TWG meeting agenda

**Comment by micaeljtoliveira** [comment]
> I have asked @micaeljtoliveira to add this to the next TWG meeting agenda.

Sure. I spend some time looking at licenses again a couple of months ago and I think I know have a much better understanding of how combining multiples licenses works.

**Comment by aidanheerdegen** [comment]
I misspoke in the meeting @micaeljtoliveira. MOM5 is LGPLv3. So not *as* viral? I guess that just means we should/could license ACCESS-OM3 as either LGPLv3 too, if we think it qualifies as code, otherwise some non-code license like CC-BY-4.0.

**Comment by micaeljtoliveira** [comment]
> MOM5 is LGPLv3. So not as viral? 

Ah, that changes things quite a lot. No, not viral at all.

> I guess that just means we should/could license ACCESS-OM3 as either LGPLv3 too, if we think it qualifies as code, otherwise some non-code license like CC-BY-4.0.

Any license compatible with all the other ones works, as none of them is viral.

Note that any code that does not **link** with the different components, like scripts, do not need to have a compatible license.

If you want, you can also include a `COPYING` file with some explanation about the fact that you are shipping some external code as git submodules and that those have their own licenses. Something like [this](https://gitlab.com/octopus-code/octopus/-/blob/main/COPYING?ref_type=heads) (although I wouldn't bother adding a copy of all the licenses in that file).

**Comment by aidanheerdegen** [comment]
I've only just noticed that the COSIMA model configs say they're covered by a [CC BY-NC-ND 4.0 license](https://creativecommons.org/licenses/by-nc-nd/4.0/legalcode):

https://github.com/COSIMA/1deg_jra55_ryf/blob/master/metadata.yaml#L16

@aekiss is there a specific reason to use a non-commercial no derivatives license? We reasoned as we're government funded it wasn't really reasonable for us to limit what uses others wanted to make of our work, including Australian commercial entities. 

Also the restriction on creating derivative works seems problematic, and would seemingly prevent ACCESS-NRI from creating derived configurations

https://creativecommons.org/2020/04/21/academic-publications-under-no-derivatives-licenses-is-misguided/

https://pitt.libguides.com/openlicensing/cc-noderivatives

**Comment by aekiss** [comment]
Hm, ok I don't recall a lot of thought going in to the CC license. I agree it seems like it should be something less restrictive.

**Comment by aidanheerdegen** [comment]
I would advocate for CC BY 4.0 as it retains the necessity to acknowledge COSIMA but places the least restrictions on derivative works but also makes the work as widely usable as possible. I think this is a worthy goal.

If you really feel strongly that the license should be unchanged in derivative works then [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/deed.en) might be a better choice.

And because more links are always good

- [Q. Which Creative Commons license should I use for my scholarly article?](https://psu.libanswers.com/faq/395000)
- [How to Use Creative Commons Licenses for Your Academic Papers](https://blog.orvium.io/creative-commons-license-academic-papers/)
- [Creative Commons Licenses: an Introduction for Researchers](https://www.aje.com/arc/creative-commons-intro/)

If you're happy to make a decision I can make PRs to update the COSIMA repos so ACCESS-NRI is consistent with COSIMA.

**Comment by aekiss** [comment]
I'm happy with CC BY 4.0

**Comment by aekiss** [comment]
all merged

**Comment by aekiss** [comment]
The MIT license is more heartfelt and beautiful than I ever realised https://twitter.com/goodside/status/1775713487529922702 (turn off mute)

**Comment by aidanheerdegen** [comment]
"the sooooftwaaaare"
